### PR TITLE
Linux bastion host

### PR DESCRIPTION
**Pull request by persiaAziz**

### CODE REVIEW

File: deploy/vm/ansible/ha_pair_playbook.yml [hunk 1]
```diff
@@ -127,3 +127,8 @@
   become: true
   roles:
     - add-AzEvents-agent
+
+- name: bastion_host playbook
```
Comment (tniek) [inline deploy/vm/ansible/ha_pair_playbook.yml]: We need to keep our code more consistent, specifically here between Windows and Linux Bastion Hosts.
For example:
- Naming: Windows bastion host is called win_bastion, Linux bastion host bastion_host => rename bastion_windows and bastion_linux, respectively
- Sequence: Windows bastion host gets created at the very beginning of playbook, Linux bastion host at the very end => have both created at the same time

File: deploy/vm/ansible/linux_bastion_host.yml [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  roles:
+          - { role: create-azure-vm, when: az_json.linux_bastion == true, azure_vm_fqdn: '{{ az_json.az_fqdn }}', public_key_path: '{{ az_json.az_public_key }}',azure_resource_group: '{{ az_json.az_resource_group }}', azure_user: '{{ az_json.az_user }}', azure_nic: linuxBastionNic}
+
+- hosts: bastion_linux_group
+  tasks:
+    - include_vars:
+        file: "{{ config_json_file }}"
+        name: az_json
+    - debug:
+            msg: " linux bastion host required: {{az_json.linux_bastion}}"
```
Comment (tniek) [inline deploy/vm/ansible/linux_bastion_host.yml]: nit: remove space before 'linux bastion host required'

File: deploy/vm/ansible/linux_bastion_host.yml [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+- hosts: localhost
+  tasks:
+    - debug:
+          msg: "linux_bastion required  :  {{az_json.linux_bastion}}"
+    - name: Create public IP address
+      azure_rm_publicipaddress:
+          resource_group: "{{ az_json.az_resource_group }}"
+          allocation_method: Dynamic
+          name: linuxbastionpip
+          domain_name: linuxbastionhost
```
Comment (tniek) [inline deploy/vm/ansible/linux_bastion_host.yml]: The naming should be consistent with the Windows bastion host

File: deploy/vm/ansible/linux_bastion_host.yml [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+      register: ip_sample
+      when: az_json.linux_bastion == true
+    - debug:
+          msg: "linux_bastion vm :{{ ip_sample.state.dns_settings.fqdn }}"
+    - name: NIC for linux bastion host
+      azure_rm_networkinterface:
+          resource_group: "{{ az_json.az_resource_group }}"
+          name: linuxBastionNic
```
Comment (tniek) [inline deploy/vm/ansible/linux_bastion_host.yml]: The naming should be consistent with the Windows bastion host

File: deploy/vm/ansible/roles/create-azure-vm/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,22 @@
+---
+- name: create Linux Bastion VM
+  vars:
+          public_key_data: "{{ lookup('file','{{ public_key_path }}') }}"
+  azure_rm_virtualmachine:
+          resource_group: "{{ azure_resource_group }}"
+          name: "linux-{{azure_resource_group}}"
+          vm_size: Standard_D8s_v3
```
Comment (tniek) [inline deploy/vm/ansible/roles/create-azure-vm/tasks/main.yml]: Bastion Host VM size should be configurable

File: deploy/vm/ansible/roles/create-azure-vm/tasks/main.yml [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+          admin_username: "{{ azure_user }}"
+          ssh_password_enabled: false
+          ssh_public_keys:
+              - path: "/home/{{ azure_user }}/.ssh/authorized_keys"
+                key_data: "{{ public_key_data }}"
+          network_interfaces: "{{ azure_nic }}"
+          started: yes
+          state: present
+          image:
```
Comment (tniek) [inline deploy/vm/ansible/roles/create-azure-vm/tasks/main.yml]: Bastion Host OS image should be configurable

File: deploy/vm/ansible/linux_bastion_host.yml [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+- hosts: bastion_linux_group
+  tasks:
+    - include_vars:
+        file: "{{ config_json_file }}"
+        name: az_json
+    - debug:
+            msg: " linux bastion host required: {{az_json.linux_bastion}}"
+
+- name: install stuffs
```
Comment (tniek) [inline deploy/vm/ansible/linux_bastion_host.yml]: Probably "install HANA Studio" would be a better description :)

File: deploy/vm/ansible/roles/create-azure-vm/tasks/main.yml [hunk 8]
```diff
@@ -0,0 +1,22 @@
+---
+- name: create Linux Bastion VM
+  vars:
+          public_key_data: "{{ lookup('file','{{ public_key_path }}') }}"
+  azure_rm_virtualmachine:
+          resource_group: "{{ azure_resource_group }}"
+          name: "linux-{{azure_resource_group}}"
```
Comment (tniek) [inline deploy/vm/ansible/roles/create-azure-vm/tasks/main.yml]: Naming of the Bastion VM should be consistent with Windows Bastion Host (should be bastion-linux)

File: deploy/vm/ansible/roles/setup-linux-hana-studio/tasks/main.yml [hunk 9]
```diff
@@ -0,0 +1,15 @@
+- name: download SAPCAR
+  get_url:
+      url: "{{sapcar_url}}"
+      dest: "/home/{{ azure_user }}/SAPCAR.EXE"
+- name: change permission of SAPCAR.EXE
+  file:
+      path: "/home/{{ azure_user }}/SAPCAR.EXE"
+      mode: 0755
+- name: download hana studio
```
Comment (tniek) [inline deploy/vm/ansible/roles/setup-linux-hana-studio/tasks/main.yml]: nit: spelling HANA Studio

File: deploy/vm/ansible/roles/setup-linux-hana-studio/tasks/main.yml [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+  get_url:
+      url: "{{hana_studio_url}}"
+      dest: "/home/{{ azure_user }}/HANA_STUDIO.SAR"
+- name: extract hana studio
```
Comment (tniek) [inline deploy/vm/ansible/roles/setup-linux-hana-studio/tasks/main.yml]: nit: spelling HANA Studio

File: deploy/vm/ansible/single_node_playbook.yml [hunk 11]
```diff
@@ -24,3 +24,8 @@
     ansible_winrm_server_cert_validation: ignore
   roles:
     - set-up-windows-bastion
+
+- name: bastion_host playbook
```
Comment (tniek) [inline deploy/vm/ansible/single_node_playbook.yml]: Probably better description would be "run playbook for Linux Bastion Host"

File: deploy/vm/ansible/ha_pair_playbook.yml [hunk 12]
```diff
@@ -127,3 +127,8 @@
...
+  vars:
+     config_json_file: ../modules/ha_pair/temp.json
```
Comment (tniek) [inline deploy/vm/ansible/ha_pair_playbook.yml]: Ugh... This is the temporary solution for passing in variables without the dynamic inventory script, right?

Comment (persiaAziz) [reply deploy/vm/ansible/ha_pair_playbook.yml]: yes it is :)

File: deploy/vm/ansible/ha_pair_playbook.yml (hunk 1 above)
Comment (persiaAziz) [reply deploy/vm/ansible/ha_pair_playbook.yml]: this is actually a comment for the play that includes the play to create Linux bastion host, bastion_host is not a variable. I will change the comment anyway

File: deploy/vm/modules/ha_pair/main.tf [hunk 13]
```diff
@@ -162,6 +162,11 @@ module "windows_bastion_host" {
   windows_bastion    = "${var.windows_bastion}"
 }
 
+resource "local_file" "write-config-to-json" {
```
Comment (pabowers) [inline deploy/vm/modules/ha_pair/main.tf]: add a comment to explain why this is here

File: deploy/vm/modules/single_node_hana/single-node-hana.tf [hunk 14]
```diff
@@ -77,3 +82,18 @@ module "configure_vm" {
   bastion_username_windows   = "${var.bastion_username_windows}"
   pw_bastion_windows         = "${var.pw_bastion_windows}"
 }
+
+resource null_resource "destroy-vm" {
```
Comment (pabowers) [inline deploy/vm/modules/single_node_hana/single-node-hana.tf]: does this get rid of that file that was created?

File: deploy/vm/ansible/roles/create-azure-vm/tasks/main.yml (hunk 5 above)
Comment (persiaAziz) [reply deploy/vm/ansible/roles/create-azure-vm/tasks/main.yml]: will create a work item

File: deploy/vm/modules/single_node_hana/single-node-hana.tf (hunk 14 above)
Comment (persiaAziz) [reply deploy/vm/modules/single_node_hana/single-node-hana.tf]: no, it does not

File: deploy/vm/ansible/roles/create-azure-vm/tasks/main.yml (hunk 6 above)
Comment (persiaAziz) [reply deploy/vm/ansible/roles/create-azure-vm/tasks/main.yml]: will create a work item

File: deploy/vm/ansible/roles/setup-linux-hana-studio/tasks/main.yml (hunk 10 above)
Comment (persiaAziz) [reply deploy/vm/ansible/roles/setup-linux-hana-studio/tasks/main.yml]: done

File: deploy/vm/ansible/single_node_playbook.yml (hunk 11 above)
Comment (persiaAziz) [reply deploy/vm/ansible/single_node_playbook.yml]: done

File: deploy/vm/modules/ha_pair/variables.tf [hunk 15]
```diff
@@ -206,7 +220,15 @@ variable "windows_bastion" {
   default     = false
 }
 
+variable "linux_bastion" {
```
[the remark this answers is not in the document]
Comment (persiaAziz) [reply deploy/vm/modules/ha_pair/variables.tf]: I see what you meant, but I really think this variable should be right after the variable "windows_bastion"

File: deploy/vm/ansible/linux_bastion_host.yml (hunk 4 above)
Comment (persiaAziz) [reply deploy/vm/ansible/linux_bastion_host.yml]: done

File: deploy/vm/ansible/linux_bastion_host.yml (hunk 2 above)
Comment (persiaAziz) [reply deploy/vm/ansible/linux_bastion_host.yml]: done

File: deploy/vm/ansible/linux_bastion_host.yml (hunk 3 above)
Comment (persiaAziz) [reply deploy/vm/ansible/linux_bastion_host.yml]: done

File: deploy/vm/modules/single_node_hana/variables.tf [hunk 16]
```diff
@@ -147,6 +156,11 @@ variable "url_sapcar_windows" {
   default     = ""
 }
 
+variable "url_sapcar_linux" {
```
[the remark this answers is not in the document]
Comment (persiaAziz) [reply deploy/vm/modules/single_node_hana/variables.tf]: will create a work item

**Comment by persiaAziz** [comment]
@tniek @pabowers I have created work items for some of the comments as they also affect some area outside the scope of this PR

File: deploy/vm/ansible/roles/create-azure-vm/tasks/main.yml (hunk 5 above)
Comment (tniek) [reply deploy/vm/ansible/roles/create-azure-vm/tasks/main.yml]: P3

File: deploy/vm/modules/single_node_hana/single-node-hana.tf (hunk 14 above)
Comment (tniek) [reply deploy/vm/modules/single_node_hana/single-node-hana.tf]: Should be the handler called when TF-destroying the VM

File: deploy/vm/modules/single_node_hana/single-node-hana.tf [hunk 17]
```diff
@@ -40,6 +40,11 @@ module "windows_bastion_host" {
   windows_bastion    = "${var.windows_bastion}"
 }
 
+resource "local_file" "write-config-to-json" {
```
[the remark this answers is not in the document]
Comment (persiaAziz) [reply deploy/vm/modules/single_node_hana/single-node-hana.tf]: done

File: deploy/vm/modules/generic_nic_and_pip/outputs.tf [hunk 18]
```diff
@@ -7,3 +7,8 @@ output "nic_id" {
   description = "The id of the network interface card will be needed to attach it to the VM."
   value       = "${azurerm_network_interface.nic.id}"
 }
+
+output "nic_name" {
+  description = "The id of the network interface"
+  value       = "${azurerm_network_interface.nic.name}"
```
[the remark this answers is not in the document]
Comment (persiaAziz) [reply deploy/vm/modules/generic_nic_and_pip/outputs.tf]: good catch

File: deploy/vm/modules/ha_pair/main.tf (hunk 13 above)
Comment (persiaAziz) [reply deploy/vm/modules/ha_pair/main.tf]: done

**Pull request merged**